<div style="text-align:center">
    <h1 style="color:green;font-size:40px">An Introduction To Python Pandas</h1
</div>

## pandas can work on data which is stored in 
    - Spreadsheets
    - Databases
## what can be done with pandas
    - data explore
    - data clean
    - data process
## In pandas a data table is called a DataFrame
   ![](https://pandas.pydata.org/docs/_images/01_table_dataframe.svg)

## Load pandas in a project
   -convention to name pandas as **pd**

In [1]:
import pandas as pd

## Create a simple dataFrame

In [2]:
df=pd.DataFrame( { "Name":["Max","John","Rana"],"Age":[23,52,21],"Sex":["male","female","male"] } )

## Each column in a DataFrame is Series in Pandas
 ![](https://pandas.pydata.org/docs/_images/01_table_series.svg)
<p>for example someone is interested in working with <span style="color:blue">Age</span>. A single column of a pandas DataFrame is a pandas Series</p>

In [4]:
df["Age"]

0    23
1    52
2    21
Name: Age, dtype: int64

## We can also create a Series from scratch.

In [6]:
marks = pd.Series([23,12,14,21,13,93,28,92],name="Marks")
marks

0    23
1    12
2    14
3    21
4    13
5    93
6    28
7    92
Name: Marks, dtype: int64


## Basic statistics of the numerical data can be shown using `describe()` method.

In [7]:
marks.describe()

count     8.000000
mean     37.000000
std      34.690468
min      12.000000
25%      13.750000
50%      22.000000
75%      44.000000
max      93.000000
Name: Marks, dtype: float64

 + REMEMBER  
      + Import the package, aka import pandas as pd

      + A table of data is stored as a pandas DataFrame

      + Each column in a DataFrame is a Series

      + You can do things by applying a method to a DataFrame or Series
      
     




## How to Read and Write tabular Data
pandas provides `read_csv()`method to read a **CSV** file and store as a **DataFrame**. It can also work with other data format. for example..
![](https://pandas.pydata.org/docs/_images/02_io_readwrite.svg)
`read_*()` method to read data from different format and `to_*()`method to write the resulting data in specific format.

In [10]:
titanic_df = pd.read_csv('datasets/tested.csv')
titanic_df

PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...   ...    ...    ...   
413                            Spector, Mr. Woolf    male   NaN      0      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.0      0      0   
415                  Saether, Mr. Simon Sivertsen    male  38.5      0      0   
416                           Ware, Mr. Frederick    male   NaN      0      0   
417                      Peter, Master. Michael J    male   NaN      1      1   

                 Ticket      Fare Cabin Embarked  
0                330911    7.8292   NaN        Q  
1                363272    7.0000   NaN        S  
2                240276    9.6875   NaN        Q  
3                315154    8.6625   NaN        S  
4               3101298   12.2875   NaN        S  
..                  ...       ...   ...      ...  
413           A.5. 3236    8.0500   NaN        S  
414            PC 17758  108.9000  C105        C  
415  SOTON/O.Q. 3101262    7.2500   NaN        S  
416              359309    8.0500   NaN        S  
417                2668   22.3583   NaN        C  

[418 rows x 12 columns]

## It is also possible to load data from the Web

First step after loading data in pandas is to check the data. Pandas `head()` methods shows by default first 5 rows. Number of row can be given as an argument

In [11]:
titanic_df.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

In [14]:
titanic_df.head(3)

PassengerId  Survived  Pclass                              Name     Sex  \
0          892         0       3                  Kelly, Mr. James    male   
1          893         1       3  Wilkes, Mrs. James (Ellen Needs)  female   
2          894         0       2         Myles, Mr. Thomas Francis    male   

    Age  SibSp  Parch  Ticket    Fare Cabin Embarked  
0  34.5      0      0  330911  7.8292   NaN        Q  
1  47.0      1      0  363272  7.0000   NaN        S  
2  62.0      0      0  240276  9.6875   NaN        Q

Last 5 rows can be shown by using `tail()` methods

In [15]:
titanic_df.tail()

PassengerId  Survived  Pclass                          Name     Sex  \
413         1305         0       3            Spector, Mr. Woolf    male   
414         1306         1       1  Oliva y Ocana, Dona. Fermina  female   
415         1307         0       3  Saether, Mr. Simon Sivertsen    male   
416         1308         0       3           Ware, Mr. Frederick    male   
417         1309         0       3      Peter, Master. Michael J    male   

      Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
413   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  39.0      0      0            PC 17758  108.9000  C105        C  
415  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416   NaN      0      0              359309    8.0500   NaN        S  
417   NaN      1      1                2668   22.3583   NaN        C

**dtypes** is an attribute in pandas **DataFrame/Series**. For each of the columns, the used data type is enlisted.
  + integers (int64)
  + floats (float64)
  + strings (object)

In [16]:
titanic_df.dtypes


PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

## Convert pandas DataFrame into a excel file

In [18]:
titanic_xlsx=titanic_df.to_excel("titanic.xlsx",sheet_name="passangers")

`info()` provides the technical summary of a **DataFrame**

In [19]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Survived     418 non-null    int64  
 2   Pclass       418 non-null    int64  
 3   Name         418 non-null    object 
 4   Sex          418 non-null    object 
 5   Age          332 non-null    float64
 6   SibSp        418 non-null    int64  
 7   Parch        418 non-null    int64  
 8   Ticket       418 non-null    object 
 9   Fare         417 non-null    float64
 10  Cabin        91 non-null     object 
 11  Embarked     418 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


+ The method `info()` provides technical information about a DataFrame, so let’s explain the output in more detail:
    + It is indeed a DataFrame.
    + There are 418 entries, i.e. 418 rows.
    + Each row has a row label (aka the index) with values ranging from 0 to 417.
    + The table has 12 columns. Most columns have a value for each of the rows (all 418 values are non-null). Some    columns do have missing values and less than 418 non-null values.
    + The columns Name, Sex, Cabin and Embarked consists of textual data (strings, aka object). The other columns are numerical data with some of them whole numbers (aka integer) and others are real numbers (aka float).
    + The kind of data (characters, integers,…) in the different columns are summarized by listing the dtypes.
    + The approximate amount of RAM used to hold the DataFrame is provided as well.



+ REMEMBER
    + Getting data in to pandas from many different file formats or data sources is supported by read_* functions.
    + Exporting data out of pandas is provided by different to_*methods.
    + The head/tail/info methods and the dtypes attribute are convenient for a first check.



In [29]:
titanic_df[titanic_df['Cabin'].isna()==True]
titanic_df["Cabin"].count()

91

## How to select a subset or specific column(s) a DataFrame 
![](https://pandas.pydata.org/docs/_images/03_subset_columns.svg)

## Select a single columns

In [35]:
ages = titanic_df['Age']
ages.head()
type(ages) # ages is also a pandas Series

pandas.core.series.Series

## another way to do the same
but this only works for a single column

In [37]:
ages = titanic_df.Age
ages.head()

0    34.5
1    47.0
2    62.0
3    27.0
4    22.0
Name: Age, dtype: float64

**shape** attribute give number of rows(entities) and columns in a **DataFrame**

In [36]:
print(ages.shape,titanic_df.shape)

(418,) (418, 12)


## Select more than one columns

In [43]:
titanic_df[['Age','Cabin','Sex']].head()
# or
list_of_cols = ['Age','Cabin','Sex']
titanic_df[list_of_cols].tail() ## same results

Age Cabin     Sex
413   NaN   NaN    male
414  39.0  C105  female
415  38.5   NaN    male
416   NaN   NaN    male
417   NaN   NaN    male

## How to filter specific rows from a DataFrame
![](https://pandas.pydata.org/docs/_images/03_subset_rows.svg)

In [50]:
titanic_df['Age']<30
less_30=titanic_df[titanic_df['Age']<30]
less_30['Age'].head()

3    27.0
4    22.0
5    14.0
7    26.0
8    18.0
Name: Age, dtype: float64

## someone is interested in The Titanic passengers from class 2 and 2

In [51]:
titanic_df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [53]:
titanic_df['Pclass'].unique()

array([3, 2, 1])

In [60]:
titanic_df[titanic_df['Pclass'].isin([3,1])].head()
## same as
titanic_df[ (titanic_df['Pclass']==3) |  (titanic_df['Pclass']==3)].head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
3          895         0       3   
4          896         1       3   
5          897         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
5                    Svensson, Mr. Johan Cervin    male  14.0      0      0   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S  
5     7538   9.2250   NaN        S

## passengers data which Age is known

In [64]:
titanic_df[titanic_df['Age'].notna()].head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

## passengers data which Cabin is known

In [66]:
titanic_df[titanic_df['Cabin'].notna()]

PassengerId  Survived  Pclass  \
12           904         1       1   
14           906         1       1   
24           916         1       1   
26           918         1       1   
28           920         0       1   
..           ...       ...     ...   
404         1296         0       1   
405         1297         0       2   
407         1299         0       1   
411         1303         1       1   
414         1306         1       1   

                                                  Name     Sex   Age  SibSp  \
12       Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0      1   
14   Chaffee, Mrs. Herbert Fuller (Carrie Constance...  female  47.0      1   
24     Ryerson, Mrs. Arthur Larned (Emily Maria Borie)  female  48.0      1   
26                        Ostby, Miss. Helene Ragnhild  female  22.0      0   
28                             Brady, Mr. John Bertram    male  41.0      0   
..                                                 ...     ...   ...    ...   
404                       Frauenthal, Mr. Isaac Gerald    male  43.0      1   
405       Nourney, Mr. Alfred (Baron von Drachstedt")"    male  20.0      0   
407                         Widener, Mr. George Dunton    male  50.0      1   
411    Minahan, Mrs. William Edward (Lillian E Thorpe)  female  37.0      1   
414                       Oliva y Ocana, Dona. Fermina  female  39.0      0   

     Parch         Ticket      Fare            Cabin Embarked  
12       0          21228   82.2667              B45        S  
14       0    W.E.P. 5734   61.1750              E31        S  
24       3       PC 17608  262.3750  B57 B59 B63 B66        C  
26       1         113509   61.9792              B36        C  
28       0         113054   30.5000              A21        S  
..     ...            ...       ...              ...      ...  
404      0          17765   27.7208              D40        C  
405      0  SC/PARIS 2166   13.8625              D38        C  
407      1         113503  211.5000              C80        C  
411      0          19928   90.0000              C78        Q  
414      0       PC 17758  108.9000             C105        C  

[91 rows x 12 columns]

## How to select specific rows and columns in a DataFrame
![](https://pandas.pydata.org/docs/_images/03_subset_columns_rows.svg)
The ***loc/iloc*** operators are required in front of the selection brackets **[rows,cols]**. When using **loc/iloc**, the part before the comma(rows) is the rows you want, and the part after the comma(cols) is the columns you want to select. <p style="color:red;">For both the part before and after the comma, you can use a single label, a list of labels, a slice of labels, a conditional expression or a colon. Using a colon specifies you want to select all rows or columns.</p>

## using a slice as rows, and list of cols 

In [83]:
titanic_df.loc[1:10,["Name",'Sex']].head(2)
## same using iloc
titanic_df.iloc[1:10,[3,4]].head(3)

Name     Sex
1  Wilkes, Mrs. James (Ellen Needs)  female
2         Myles, Mr. Thomas Francis    male
3                  Wirz, Mr. Albert    male

## find names of Passengers  older than 35 using logical expression as rows

In [100]:
titanic_df.loc[titanic_df['Age']>35,['Name']].head()
## same using iloc
titanic_df.iloc[ list(titanic_df['Age']>35),[3,4]  ].head()


Name     Sex
1                    Wilkes, Mrs. James (Ellen Needs)  female
2                           Myles, Mr. Thomas Francis    male
11                         Jones, Mr. Charles Cresson    male
13                               Howard, Mr. Benjamin    male
14  Chaffee, Mrs. Herbert Fuller (Carrie Constance...  female

<p style="color:red" >When selecting specific rows and/or columns with <span style="font-size:bold">loc or iloc, new values can be assigned to the selected data. For example, to assign the name anonymous to the first 3 elements of the third column:</p>

In [107]:
titanic_df.iloc[0:10,3]="Rana"
titanic_df[["Name", "Sex"]].head()

Name     Sex
0  Rana    male
1  Rana  female
2  Rana    male
3  Rana    male
4  Rana  female

## difference between loc and iloc
| **loc** | **iloc** |
|-----|------|
|  basically based on labels | based on indexing |


 + REMEMBER

   + When selecting subsets of data, square brackets [] are used.
  
   +  Inside these brackets, you can use a single column/row label, a list of column/row labels, a slice of labels, a conditional expression or a colon.

   + Select specific rows and/or columns using loc when using the row and column names.

   + Select specific rows and/or columns using iloc when using the positions in the table.

   + You can assign new values to a selection based on loc/iloc.

